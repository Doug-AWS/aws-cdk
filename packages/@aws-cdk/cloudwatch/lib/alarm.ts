import { Arn, Construct, Token } from '@aws-cdk/core';
import { cloudwatch } from '@aws-cdk/resources';
import { Metric } from './metric';

/**
 * Properties for Alarms
 */
export interface AlarmProps {
    /**
     * The metric to add the alarm on
     */
    metric: Metric;

    /**
     * Name of the alarm
     *
     * @default Automatically generated name
     */
    alarmName?: string;

    /**
     * Description for the alarm
     *
     * @default No description
     */
    alarmDescription?: string;

    /**
     * Comparison to use to check if metric is breaching
     *
     * @default GreaterThanOrEqualToThreshold
     */
    comparisonOperator?: ComparisonOperator;

    /**
     * The value against which the specified statistic is compared.
     */
    threshold: number;

    /**
     * The number of periods over which data is compared to the specified threshold.
     */
    evaluationPeriods: number;

    /**
     * Number of datapoints that must be breaching to trigger the alarm
     *
     * This is used only if you are setting an "M out of N" alarm. In that case, this value is the M.
     *
     * @default Not an "M out of N" alarm.
     */
    datapointsToAlarm?: number;

    /**
     * Specifies whether to evaluate the data and potentially change the alarm state if there are too few data points to be statistically significant.
     *
     * Used only for alarms that are based on percentiles.
     */
    evaluateLowSampleCountPercentile?: string;

    /**
     * Sets how this alarm is to handle missing data points.
     *
     * @default TreatMissingData.Missing
     */
    treatMissingData?: TreatMissingData;

    /**
     * Whether the actions for this alarm are enabled
     *
     * @default true
     */
    actionsEnabled?: boolean;
}

/**
 * Comparison operator for evaluating alarms
 */
export enum ComparisonOperator {
    GreaterThanOrEqualToThreshold = 'GreaterThanOrEqualToThreshold',
    GreaterThanThreshold = 'GreaterThanThreshold',
    LessThanThreshold = 'LessThanThreshold',
    LessThanOrEqualToThreshold = 'LessThanOrEqualToThreshold',
}

/**
 * Specify how missing data points are treated during alarm evaluation
 */
export enum TreatMissingData {
    /**
     * Missing data points are treated as breaching the threshold
     */
    Breaching = 'breaching',

    /**
     * Missing data points are treated as being within the threshold
     */
    NotBreaching = 'notBreaching',

    /**
     * The current alarm state is maintained
     */
    Ignore = 'ignore',

    /**
     * The alarm does not consider missing data points when evaluating whether to change state
     */
    Missing = 'missing'
}

/**
 * An alarm on a CloudWatch metric
 */
export class Alarm extends Construct {
    /**
     * ARN of this alarm
     */
    public readonly alarmArn: AlarmArn;

    /**
     * The metric object this alarm was based on
     */
    public readonly metric: Metric;

    private alarmActions?: Arn[];
    private insufficientDataActions?: Arn[];
    private okActions?: Arn[];

    constructor(parent: Construct, name: string, props: AlarmProps) {
        super(parent, name);

        const alarm = new cloudwatch.AlarmResource(this, 'Resource', {
            actionsEnabled: props.actionsEnabled,
            alarmActions: new Token(() => this.alarmActions),
            alarmDescription: props.alarmDescription,
            alarmName: props.alarmName,
            comparisonOperator: props.comparisonOperator || ComparisonOperator.GreaterThanOrEqualToThreshold,
            evaluateLowSampleCountPercentile: props.evaluateLowSampleCountPercentile,
            evaluationPeriods: props.evaluationPeriods,
            insufficientDataActions: new Token(() => this.insufficientDataActions),
            okActions: new Token(() => this.okActions),
            threshold: props.threshold,
            treatMissingData: props.treatMissingData,
            ...props.metric.toAlarmJson()
        });

        this.alarmArn = alarm.alarmArn;
        this.metric = props.metric;
    }

    /**
     * Trigger this action if the alarm fires
     *
     * Typically the ARN of an SNS topic or ARN of an AutoScaling policy.
     */
    public onAlarm(arn: Arn) {
        if (this.alarmActions === undefined) {
            this.alarmActions = [];
        }

        this.alarmActions.push(arn);
    }

    /**
     * Trigger this action if there is insufficient data to evaluate the alarm
     *
     * Typically the ARN of an SNS topic or ARN of an AutoScaling policy.
     */
    public onInsufficientData(arn: Arn) {
        if (this.insufficientDataActions === undefined) {
            this.insufficientDataActions = [];
        }

        this.insufficientDataActions.push(arn);
    }

    /**
     * Trigger this action if the alarm returns from breaching state into ok state
     *
     * Typically the ARN of an SNS topic or ARN of an AutoScaling policy.
     */
    public onOk(arn: Arn) {
        if (this.okActions === undefined) {
            this.okActions = [];
        }

        this.okActions.push(arn);
    }
}

/**
 * The ARN of an Alarm
 */
export class AlarmArn extends Arn {
}