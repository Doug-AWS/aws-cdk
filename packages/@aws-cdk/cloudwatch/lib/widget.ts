/**
 * The width of the grid we're filling
 */
export const GRID_WIDTH = 24;

/**
 * A single dashboard widget
 */
export interface IWidget {
    /**
     * The width of the widget
     *
     * Will only ever be queried after 'determineSize' has been called.
     */
    readonly width: number;

    /**
     * The height of the widget
     *
     * Will only ever be queried after 'determineSize' has been called.
     */
    readonly height: number;

    /**
     * Place the widget at a given position
     */
    position(x: number, y: number): void;

    /**
     * Return the widget JSON for use in the dashboard
     */
    toJson(): any[];
}

/**
 * A real CloudWatch widget that has its own fixed size and remembers its position
 *
 * This is in contrast to other widgets which exist for layout purposes.
 */
export abstract class ConcreteWidget implements IWidget {
    public readonly width: number;
    public readonly height: number;
    protected x?: number;
    protected y?: number;

    constructor(width: number, height: number) {
        this.width = width;
        this.height = height;

        if (this.width > GRID_WIDTH) {
            throw new Error(`Widget is too wide, max ${GRID_WIDTH} units allowed`);
        }
    }

    public position(x: number, y: number): void {
        this.x = x;
        this.y = y;
    }

    public abstract toJson(): any[];
}