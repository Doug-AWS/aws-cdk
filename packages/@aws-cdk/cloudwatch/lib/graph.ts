import { AwsRegion, Token } from "@aws-cdk/core";
import { Alarm } from "./alarm";
import { Metric } from "./metric";
import { ConcreteWidget } from "./widget";

/**
 * An AWS region
 */
export class Region extends Token {
}

/**
 * Basic properties for widgets that display metrics
 */
export interface MetricWidgetProps {
    /**
     * Title for the graph
     */
    title?: string;

    /**
     * The region the metrics of this graph should be taken from
     *
     * @default Current region
     */
    region?: Region;

    /**
     * Width of the widget, in a grid of 24 units wide
     *
     * @default 6
     */
    width?: number;

    /**
     * Height of the widget
     *
     * @default 6
     */
    height?: number;
}

/**
 * Properties for an AlarmWidget
 */
export interface AlarmWidgetProps extends MetricWidgetProps {
    /**
     * The alarm to show
     */
    alarm: Alarm;

    /**
     * Minimum of left Y axis
     *
     * @default 0
     */
    leftAxisMin?: number;

    /**
     * Maximum of left Y axis
     *
     * @default Automatic
     */
    leftAxisMax?: number;
}

/**
 * Display the metric associated with an alarm, including the alarm line
 */
export class AlarmWidget extends ConcreteWidget {
    private readonly props: AlarmWidgetProps;

    constructor(props: AlarmWidgetProps) {
        super(props.width || 6, props.height || 6);
        this.props = props;
    }

    public toJson(): any[] {
        return [{
            type: 'metric',
            width: this.width,
            height: this.height,
            x: this.x,
            y: this.y,
            properties: {
                view: 'timeSeries',
                title: this.props.title,
                region: this.props.region || new AwsRegion(),
                annotations: {
                    alarms: [this.props.alarm.alarmArn]
                },
                yAxis: {
                    left: {
                        min: this.props.leftAxisMin,
                        max: this.props.leftAxisMax
                    }
                }
            }
        }];
    }
}

/**
 * Properties for a GraphWidget
 */
export interface GraphWidgetProps extends MetricWidgetProps {
    /**
     * Metrics to display on left Y axis
     */
    left?: Metric[];

    /**
     * Metrics to display on right Y axis
     */
    right?: Metric[];

    /**
     * Annotations for the left Y axis
     */
    leftAnnotations?: HorizontalAnnotation[];

    /**
     * Annotations for the right Y axis
     */
    rightAnnotations?: HorizontalAnnotation[];

    /**
     * Whether the graph should be shown as stacked lines
     */
    stacked?: boolean;

    /**
     * Minimum of left Y axis
     *
     * @default 0
     */
    leftAxisMin?: number;

    /**
     * Maximum of left Y axis
     *
     * @default Automatic
     */
    leftAxisMax?: number;

    /**
     * Minimum of right Y axis
     *
     * @default 0
     */
    rightAxisMin?: number;

    /**
     * Maximum of right Y axis
     *
     * @default Automatic
     */
    rightAxisMax?: number;
}

/**
 * A dashboard widget that displays MarkDown
 */
export class GraphWidget extends ConcreteWidget {
    private readonly props: GraphWidgetProps;

    constructor(props: GraphWidgetProps) {
        super(props.width || 6, props.height || 6);
        this.props = props;
    }

    public toJson(): any[] {
        return [{
            type: 'metric',
            width: this.width,
            height: this.height,
            x: this.x,
            y: this.y,
            properties: {
                view: 'timeSeries',
                title: this.props.title,
                region: this.props.region || new AwsRegion(),
                metrics: (this.props.left || []).map(m => m.toGraphJson('left')).concat(
                         (this.props.right || []).map(m => m.toGraphJson('right'))),
                annotations: {
                    horizontal: (this.props.leftAnnotations || []).map(mapAnnotation('left')).concat(
                                (this.props.leftAnnotations || []).map(mapAnnotation('right')))
                },
                yAxis: {
                    left: {
                        min: this.props.leftAxisMin,
                        max: this.props.leftAxisMax
                    },
                    right: {
                        min: this.props.rightAxisMin,
                        max: this.props.rightAxisMax
                    },
                }
            }
        }];
    }
}

/**
 * Properties for a SingleValueWidget
 */
export interface SingleValueWidgetProps extends MetricWidgetProps {
    /**
     * Metrics to display
     */
    metrics: Metric[];
}

/**
 * A dashboard widget that displays the most recent value for every metric
 */
export class SingleValueWidget extends ConcreteWidget {
    private readonly props: SingleValueWidgetProps;

    constructor(props: SingleValueWidgetProps) {
        super(props.width || 6, props.height || 6);
        this.props = props;
    }

    public toJson(): any[] {
        return [{
            type: 'metric',
            width: this.width,
            height: this.height,
            x: this.x,
            y: this.y,
            properties: {
                view: 'singleValue',
                title: this.props.title,
                region: this.props.region || new AwsRegion(),
                metrics: this.props.metrics.map(m => m.toGraphJson('left'))
            }
        }];
    }
}

/**
 * Horizontal annotation to be added to a graph
 */
export interface HorizontalAnnotation {
    /**
     * The value of the annotation
     */
    value: number;

    /**
     * Label for the annotation
     *
     * @default No label
     */
    label?: string;

    /**
     * Hex color code to be used for the annotation
     *
     * @default Automatic color
     */
    color?: string;

    /**
     * Add shading above or below the annotation
     *
     * @default No shading
     */
    fill?: Shading;

    /**
     * Whether the annotation is visible
     *
     * @default true
     */
    visible?: boolean;
}

export enum Shading {
    /**
     * Don't add shading
     */
    None = 'none',

    /**
     * Add shading above the annotation
     */
    Above = 'above',

    /**
     * Add shading below the annotation
     */
    Below = 'below'
}

function mapAnnotation(yAxis: string): ((x: HorizontalAnnotation) => any) {
    return (a: HorizontalAnnotation) => {
        return { ...a, yAxis };
    };
}