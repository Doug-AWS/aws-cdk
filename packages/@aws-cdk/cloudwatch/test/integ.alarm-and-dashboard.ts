// Integration test to deploy some resources, create an alarm on it and create a dashboard.
//
// Because literally every other library is going to depend on @aws-cdk/cloudwatch, we drop down
// to the very lowest level to create CloudFormation resources by hand, without even generated
// library support.

import { App, Resource, Stack } from '@aws-cdk/core';
import { AlarmWidget, Dashboard, Metric } from '../lib';

const app = new App(process.argv);

const stack = new Stack(app, `aws-cdk-cloudwatch`);

const queue = new Resource(stack, 'queue', { type: 'AWS::SQS::Queue' });

const metric = new Metric({
    namespace: 'AWS/SQS',
    metricName: 'ApproximateNumberOfMessagesVisible',
    dimensions: { QueueName: queue.getAtt('QueueName') }
});

const alarm = metric.newAlarm(stack, 'Alarm', {
    threshold: 100,
    evaluationPeriods: 3
});

const dashboard = new Dashboard(stack, 'Dash');
dashboard.add(new AlarmWidget({
    title: 'Messages in queue',
    alarm,
}));

process.stdout.write(app.run());
