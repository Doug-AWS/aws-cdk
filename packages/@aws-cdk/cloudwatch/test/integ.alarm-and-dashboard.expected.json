{
  "Resources": {
    "queue": {
      "Type": "AWS::SQS::Queue"
    },
    "Alarm7103F465": {
      "Type": "AWS::CloudWatch::Alarm",
      "Properties": {
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "queue",
                "QueueName"
              ]
            }
          }
        ],
        "EvaluationPeriods": 3,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 100
      }
    },
    "DashCCD7F836": {
      "Type": "AWS::CloudWatch::Dashboard",
      "Properties": {
        "DashboardBody": {
          "Fn::Sub": [
            "{\"widgets\":[{\"type\":\"metric\",\"width\":6,\"height\":6,\"x\":0,\"y\":0,\"properties\":{\"view\":\"timeSeries\",\"title\":\"Messages in queue\",\"region\":\"${ref0}\",\"annotations\":{\"alarms\":[\"${ref1}\"]},\"yAxis\":{\"left\":{\"min\":0}}}}]}",
            {
              "ref0": {
                "Ref": "AWS::Region"
              },
              "ref1": {
                "Fn::GetAtt": [
                  "Alarm7103F465",
                  "Arn"
                ]
              }
            }
          ]
        }
      }
    }
  }
}